File: 2026-1/sesion-de-clase-03/dotnet/seguro-vehicular/Program.cs
using Coberturas;
using Seguros;

public class Program
{
    public static void Main(string[] args)
    {
        Vehiculo vehiculo = new()
        {
            Antiguedad = 2,
            Marca = Marca.Toyota
        };

        Console.WriteLine("===== SEGUROS PREDEFINIDOS =====");

        ICotizable seguro = new SeguroBasico(vehiculo);
        Console.WriteLine(seguro);

        seguro = new SeguroBronce(vehiculo);
        Console.WriteLine(seguro);

        seguro = new SeguroPlata(vehiculo);
        Console.WriteLine(seguro);

        seguro = new SeguroOro(vehiculo);
        Console.WriteLine(seguro);

        /* Implementación utilizando el Patrón Decorador 
            * para extender funcionalidades de manera flexible */
        Console.WriteLine("\n===== SEGUROS DECORADOS (Patrón Decorador) =====");

        seguro = new SeguroBasico(vehiculo);
        Console.WriteLine(seguro);

        seguro = new SeguroBasico(vehiculo)
            .ConCoberturaContraRobos();
        Console.WriteLine(seguro);

        seguro = new SeguroBasico(vehiculo)
            .ConCoberturaContraRobos()
            .ConAsistenciaVial();
        Console.WriteLine(seguro);

        seguro = new SeguroBasico(vehiculo)
            .ConCoberturaContraRobos()
            .ConAsistenciaVial()
            .ConCoberturaContraDesastresNaturales();
        Console.WriteLine(seguro);

        seguro = new SeguroBasico(vehiculo)
            .ConCoberturaContraRobos()
            .ConCoberturaContraDesastresNaturales();
        Console.WriteLine(seguro);
    }
}

File: 2026-1/sesion-de-clase-03/java/seguro-vehicular/Program.java

public class Program {

    public static void main(String[] args) {
        Vehiculo vehiculo = new Vehiculo(Marca.Toyota, 2);

        System.out.println("===== SEGUROS PREDEFINIDOS =====");

        Cotizable seguro = new SeguroBasico(vehiculo);
        System.out.println(seguro);

        seguro = new SeguroBronce(vehiculo);
        System.out.println(seguro);

        seguro = new SeguroPlata(vehiculo);
        System.out.println(seguro);

        seguro = new SeguroOro(vehiculo);
        System.out.println(seguro);

        /* Implementación utilizando el Patrón Decorador 
            * para extender funcionalidades de manera flexible */
        System.out.println("\n===== SEGUROS DECORADOS (Patrón Decorador) =====");

        seguro = new SeguroBasico(vehiculo);
        System.out.println(seguro);

        seguro = new CoberturaContraRobos(new SeguroBasico(vehiculo));
        System.out.println(seguro);

        seguro = new CoberturaAsistenciaVial(new CoberturaContraRobos(new SeguroBasico(vehiculo)));
        System.out.println(seguro);

        seguro = new CoberturaContraDesastresNaturales(new CoberturaAsistenciaVial(new CoberturaContraRobos(new SeguroBasico(vehiculo))));
        System.out.println(seguro);

        seguro = new CoberturaContraDesastresNaturales(new CoberturaContraRobos(new SeguroBasico(vehiculo)));
        System.out.println(seguro);
    }
}
